### PR TITLE
feat(framework): refactor as_ivy_dev and as_native to check for available device on local hardware

### PR DESCRIPTION
**Pull request by abdulasiraj**
# PR Description

## Related Issue

Close #

## Checklist

- [ ] Did you add a function?
- [ ] Did you add the tests?
- [ ] Did you run your tests and are your tests passing?
- [ ] Did pre-commit not fail on any check?
- [ ] Did you follow the steps we provided?

### CODE REVIEW

**Comment by abdulasiraj** [comment]
If you're on the GPU. please test locally to confirm things. thanks!

File: ivy/functional/backends/jax/device.py [hunk 1]
```diff
@@ -76,27 +76,104 @@ def _to_device(x, device=None):
     return x
 
 
+def is_valid_device(device_platform, device_id, /):
```
Comment (vedpatwardhan) [inline ivy/functional/backends/jax/device.py:79]: let's make this a private helper

File: ivy/functional/backends/jax/device.py [hunk 2]
```diff
@@ -76,27 +76,104 @@ def _to_device(x, device=None):
...
+    return device_platform in ["gpu", "tpu"] and device_id in range(
+        0, jax.device_count(device_platform)
+    )
+
+
 def as_ivy_dev(device, /):
-    if isinstance(device, str):
-        return ivy.Device(device)
+    """
+    Convert a JAX device to an ivy Device.
```
Comment (vedpatwardhan) [inline ivy/functional/backends/jax/device.py:87]: we shouldn't need a docstring to a backend-specific function. Given that all backends should have the same behaviour, adding the docstring to `as_ivy_dev` in `ivy/functional/ivy/device.py` should be sufficient

File: ivy/functional/backends/jax/device.py [hunk 3]
```diff
@@ -76,27 +76,104 @@ def _to_device(x, device=None):
...
+    if is_valid_device(p, dev_id):
+        return ivy.Device(p + ":" + str(dev_id))
+    else:
+        return ivy.Device(p + ":" + str(0))
 
 
 def as_native_dev(device, /):
-    if not isinstance(device, str):
+    """
+    Convert an ivy Device to a JAX device.
```
Comment (vedpatwardhan) [inline ivy/functional/backends/jax/device.py:139]: same here, let's remove the docstring

File: ivy/functional/backends/jax/device.py [hunk 4]
```diff
@@ -76,27 +76,104 @@ def _to_device(x, device=None):
...
+    ):
         return device
     dev_split = ivy.Device(device).split(":")
     device = dev_split[0]
     if len(dev_split) > 1:
         idx = int(dev_split[1])
     else:
         idx = 0
-    return jax.devices(device)[idx]
+
```
Comment (vedpatwardhan) [inline ivy/functional/backends/jax/device.py:169]: We could change this implementation to something like this right?
```
    if isinstance(device, jaxlib.xla_extension.Device):
        return device
    device_platform = device
    if isinstance(device, str):
        device = ivy.Device(device)
        device_platform, device_id = (device[0:3], int(device[4:]))
    if device_platform in [None, "cpu"]:
        return jax.devices("cpu")[0]
    if is_valid_device(device_platform, device_id):
        return jax.devices(device_platform)[device_id]
    else:
        return jax.devices(device_platform)[0]
```

File: ivy/functional/ivy/device.py [hunk 5]
```diff
@@ -387,6 +387,10 @@ def dev(
 # Conversions
 
 
+def get_all_native_devices():
```
Comment (vedpatwardhan) [inline ivy/functional/ivy/device.py:390]: Not sure about the purpose of this function, could you please elaborate?

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]
Hey @abdulasiraj, just requested a few changes. Thanks 😄

File: ivy/functional/ivy/device.py (hunk 5 above)
Comment (abdulasiraj) [reply ivy/functional/ivy/device.py:390]: was a local testing function. pushed somehow.

File: ivy/functional/backends/jax/device.py (hunk 4 above)
Comment (abdulasiraj) [reply ivy/functional/backends/jax/device.py:169]: yeah,

File: ivy/functional/backends/jax/device.py (hunk 3 above)
Comment (abdulasiraj) [reply ivy/functional/backends/jax/device.py:139]: sure

**Comment by abdulasiraj** [comment]
Hey @vedpatwardhan,
this is final now. I tried to generalize as much as i can. please have an other look. 
thanks!

File: ivy/functional/backends/jax/device.py [hunk 6]
```diff
@@ -60,43 +60,28 @@ def to_device(
...
     if device is not None:
-        cur_dev = as_native_dev(dev(x))
+        cur_dev = ivy.as_native_dev(dev(x))
         if cur_dev != device:
-            x = jax.device_put(x, as_native_dev(device))
+            x = jax.device_put(x, ivy.as_native_dev(device))
     return x
 
 
-def as_ivy_dev(device, /):
```
Comment (vedpatwardhan) [inline ivy/functional/backends/jax/device.py:79]: Seems like you've deleted the `as_ivy_dev` and `as_native_dev` functions which were to be refactored 😅
Could you please add them back and make the changes as suggested in the previous review? Thanks @abdulasiraj!

Comment (vedpatwardhan) [reply ivy/functional/backends/jax/device.py:79]: same applies to the other backends

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]

File: ivy/functional/backends/jax/device.py (hunk 6 above)
Comment (abdulasiraj) [reply ivy/functional/backends/jax/device.py:79]: done

**Comment by abdulasiraj** [comment]
Hey @vedpatwardhan,
it final now hopefully. mind taking another look?
thanks!

File: ivy/__init__.py [hunk 7]
```diff
@@ -103,6 +103,8 @@ def __new__(cls, dev_str):
             ivy.utils.assertions.check_elem_in_list(dev_str[0:3], ["gpu", "tpu", "cpu"])
             if dev_str != "cpu":
                 # ivy.assertions.check_equal(dev_str[3], ":")
+                if len(dev_str) == 3:
```
Comment (vedpatwardhan) [inline ivy/__init__.py:106]: let's remove this, `torch` also allows not having to pass the `id` 😅
```
x = torch.tensor([1., 2., 3.], device="cuda")
```
So basically just like `gpu:100` falls back to `gpu:0`, `gpu` would also be read as `gpu:0`

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]
Just a small change, rest seems good to merge. Thanks @abdulasiraj 😄

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by Ishticode** [comment]
Closing this due to inactivity after changes were requested for over 6 months plus its been labelled stale for some time. Please feel free to reopen if you would like to continue working on this. Thank you :)
